add signature VerifySigning for dotnet tools

JL03-Yue opened the pull request:
#37469 

Add signature verifications for NuGet Tools

JL03-Yue committed:
add signature VerifySigning for dotnet tools and add tests

JL03-Yue committed:
fix tests

JL03-Yue committed:
fix tests

dsplaisted commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -86,7 +86,8 @@ internal class NuGetPackageDownloader : INuGetPackageDownloader
             DirectoryPath? downloadFolder = null,
-            PackageSourceMapping packageSourceMapping = null)
+            PackageSourceMapping packageSourceMapping = null,
+            bool verifySigning = true)
```
What's the difference between the `verifySigning` parameter to this method and the `verifySignatures` constructor parameter?  They sound like the same thing.  I think either we should just have one setting (probably the constructor parameter), or we need very clear comments explaining why we need two different values and what the difference between them is.

JL03-Yue replied:
Thank you very much for the review. Apologize for the duplicated definition of `verifySignatures`. The duplicated is removed and the settings is set in the constructer of `NuGetPackageDownloader`.

dsplaisted commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,18 +161,48 @@ private void VerifySigning(string nupkgPath)
 
+        private static bool NuGetVerify(FilePath nupkgToVerify, out string commandOutput)
+        {
+            var args = new[] { "verify", "--all", nupkgToVerify.Value };
+            var command = new DotNetCommandFactory(alwaysRunOutOfProc: true)
+                .Create("nuget", args);
```
Did you get feedback from the NuGet team on how to do the signature verification? Is starting a new process the only way to do this?  Are there NuGet APIs we could call in-process instead?

JL03-Yue replied:
Thank you very much. I talked to @Damon and he suggests that calling the [nuget verify command](https://learn.microsoft.com/dotnet/core/tools/dotnet-nuget-verify) is the preferred approach because it is easier to maintain over the long term. "It would be possible to invoke the same functionality via API, but any API surface that's exposed today was less the result of an intentional, well crafted API design and more the result of a "let's make everything public by default" team policy. " Overall, it seems that calling the nuget verify command is valid and is a good approach.

dsplaisted reviewed (commented):
Some scenarios to test (manually):

- Can you install a non-Microsoft tool from NuGet.org?
- Can you install a locally-built tool (which won't be signed at all)?

JL03-Yue committed:
remove duplicate definition of verifysignature

JL03-Yue committed:
fix tests

JL03-Yue committed:
fix builds

JL03-Yue committed:
fix tests

JL03-Yue commented:
Thank you very much for the review!

For the two manual test questions:

- Can you install a non-Microsoft tool from NuGet.org?

Yes, for example, `Cake.tool` can be installed. It sounds to me that the tool can have various signings, such as package signs and organization signs. 

- Can you install a locally-built tool (which won't be signed at all)?

No, from manual test, locally built tools cannot be installed because the signing is not validated. I think this is not the expected behavior and will need a work-around about this.

dsplaisted commented:
> No, from manual test, locally built tools cannot be installed because the signing is not validated. I think this is not the expected behavior and will need a work-around about this.

Yes, I think we would want to use the same type of logic that NuGet uses for PackageReferences to determine whether we need to verify whether the package is signed.

kartheekp-ms commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,18 +154,48 @@ private void VerifySigning(string nupkgPath)
             {
-                if (!_firstPartyNuGetPackageSigningVerifier.Verify(new FilePath(nupkgPath),
-                    out string commandOutput))
+                if(RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Given that NuGet package signature verification is enabled by default on Linux in .NET 8 https://github.com/dotnet/sdk/pull/31868 PR, I think we could enable signature verification for .NET Tools NuGet packages on Linux also.

cc @dtivel

dtivel replied:
Can the default behavior be the same as https://github.com/dotnet/sdk/pull/31868/files, driven by the same environment variable?

kartheekp-ms commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,18 +154,48 @@ private void VerifySigning(string nupkgPath)
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            if (!_isNuGetTool)
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Same as https://github.com/dotnet/sdk/pull/39268/files#r1586876370

kartheekp-ms commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,18 +154,48 @@ private void VerifySigning(string nupkgPath)
                 return;
```
I am not sure if this is the correct location to add the comment. It could be possible that I don't have a good understanding of the code in this repo. I think we should pass `verifySignatures to true` when trying to execute the dotnet tool install command.

https://github.com/JL03-Yue/sdk/blob/684d26597df5b1e06a799ce7481d97e6670fdd79/src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs#L84

If we don't pass `verifySignatures = true` then I think the if block will return immediately there by skipping signature verification.

```cs
nugetPackageDownloader ??= new NuGetPackageDownloader(tempDir, verboseLogger: new NullLogger(), restoreActionConfig: restoreAction, verbosityOptions: _verbosity);
```

kartheekp-ms replied:
I am not sure if this is the correct location to add the comment. It could be possible that I don't have a good understanding of the code in this repo. I think we should set `verifySignatures` to `true` when trying to execute the `dotnet tool install` command.

https://github.com/JL03-Yue/sdk/blob/684d26597df5b1e06a799ce7481d97e6670fdd79/src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs#L84

If we don't set `verifySignatures = true`, then I think this `if` block will cause the control to return immediately, thereby skipping signature verification.

```cs
nugetPackageDownloader ??= new NuGetPackageDownloader(tempDir, verboseLogger: new NullLogger(), restoreActionConfig: restoreAction, verbosityOptions: _verbosity);
```

kartheekp-ms commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,18 +154,48 @@ private void VerifySigning(string nupkgPath)
             }
         }
 
+        private static bool NuGetVerify(FilePath nupkgToVerify, out string commandOutput)
+        {
+            var args = new[] { "verify", "--all", nupkgToVerify.Value };
+            var command = new DotNetCommandFactory(alwaysRunOutOfProc: true)
+                .Create("nuget", args);
+
+            var commandResult = command.CaptureStdOut().Execute();
+            commandOutput = commandResult.StdOut + Environment.NewLine + commandResult.StdErr;
+            return commandResult.ExitCode == 0;
+        }
+
```
Can we leverage the existing logic instead of code duplication?https://github.com/dotnet/sdk/blob/45443375ec4a612d43b6888bb1c08770565614b8/src/Cli/dotnet/NugetPackageDownloader/FirstPartyNuGetPackageSigningVerifier.cs#L36

Forgind committed:
Resolve comments

Forgind committed:
Change that looks good but isn't right

Forgind committed:
Fix NuGet signing check

Forgind committed:
Merge branch 'main' into tool-update-signature-verification

Forgind committed:
Change name

Forgind committed:
Add paramName:

Forgind commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,14 +154,22 @@ private void VerifySigning(string nupkgPath)
             }
 
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            if (await repository.GetResourceAsync<RepositorySignatureResource>().ConfigureAwait(false) is var resource &&
```
```suggestion
            if (await repository?.GetResourceAsync<RepositorySignatureResource>().ConfigureAwait(false) is var resource &&
```

Forgind committed:
Update src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

Forgind commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,14 +154,22 @@ private void VerifySigning(string nupkgPath)
             }
 
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            if (await repository?.GetResourceAsync<RepositorySignatureResource>().ConfigureAwait(false) is var resource &&
```
```suggestion
            if (repository is not null && await repository.GetResourceAsync<RepositorySignatureResource>().ConfigureAwait(false) is var resource &&
```

Forgind committed:
Update src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

Forgind committed:
little tweak

Forgind committed:
Delete test that can no longer work

This test can no longer work. The sign check only happens when a nuget package source says the nuget package should be signed, and the only way to make that happen in a test is to download the package from an https resource, which generally isn't a good idea in the middle of a test regardless and certainly isn't a good idea if we need to find a package that should be signed but isn't (i.e., may be malicious). I propose replacing this test with manual tests.

Forgind committed:
Add messages

marcpopMSFT commented:
@Forgind is this now ready for review (modulo the remaining failing leg). Added @joeloff to review.

Forgind committed:
Make test skip sign check

Forgind committed:
Make other test pass (and merge into theory)

dsplaisted commented on src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs:
```diff
@@ -157,15 +157,25 @@ private void VerifySigning(string nupkgPath)
-                if (!_firstPartyNuGetPackageSigningVerifier.Verify(new FilePath(nupkgPath),
-                    out string commandOutput))
+                string commandOutput;
+                if ((!_shouldUsePackageSourceMapping && !_firstPartyNuGetPackageSigningVerifier.Verify(new FilePath(nupkgPath), out commandOutput)) ||
+                    (_shouldUsePackageSourceMapping && !FirstPartyNuGetPackageSigningVerifier.NuGetVerify(new FilePath(nupkgPath), out commandOutput, _currentWorkingDirectory)))
```
What's the difference between `_firstPartyNuGetPackageSigningVerifier.Verify` and `FirstPartyNuGetPackageSigningVerifier.NuGetVerify`?  That's not obvious to me and it seems like a comment explaining would be helpful.

Forgind replied:
They certainly look the same. I can add a comment. It's whether or not it should additionally check that it's a first-party binary or not.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs:
```diff
@@ -89,7 +94,7 @@ public ToolInstallGlobalOrToolPathCommand(
                 IgnoreFailedSources: parseResult.GetValue(ToolCommandRestorePassThroughOptions.IgnoreFailedSourcesOption),
                 Interactive: parseResult.GetValue(ToolCommandRestorePassThroughOptions.InteractiveRestoreOption));
-            nugetPackageDownloader ??= new NuGetPackageDownloader(tempDir, verboseLogger: new NullLogger(), restoreActionConfig: restoreAction, verbosityOptions: _verbosity);
+            nugetPackageDownloader ??= new NuGetPackageDownloader(tempDir, verboseLogger: new NullLogger(), restoreActionConfig: restoreAction, verbosityOptions: _verbosity, verifySignatures: true);
```
It looks like `_verifySignatures` isn't being used here.  Shouldn't it be?

```suggestion
            nugetPackageDownloader ??= new NuGetPackageDownloader(tempDir, verboseLogger: new NullLogger(), restoreActionConfig: restoreAction, verbosityOptions: _verbosity, verifySignatures: _verifySignatures ?? true);
```

Forgind replied:
This NuGetPackageDownloader is only used to find the app host, so I didn't think I needed to change this, but when I dug a little deeper just now, it does download the app host, so I think that if someone has a feed with a malicious version of the app host that returns faster than nuget.org, that could be bad, so I think you're right. Good catch.

Forgind replied:
Ok. Note that verifySignatures was not available here before the test change, so if the consensus is that it is needed for that nugetPackageDownloader, then we need to keep (at least part of) the test changes as well.

dsplaisted commented on test/Microsoft.DotNet.PackageInstall.Tests/ToolPackageDownloaderTests.cs:
```diff
@@ -54,7 +54,8 @@ public void GivenNugetConfigInstallSucceeds(bool testMockBehaviorIsInSync)
                 targetFramework: _testTargetframework,
-                isGlobalTool: true);
+                isGlobalTool: true,
+                verifySignatures: false);
```
Do we need all these test changes to specify verifySignatures = false?  It seems to me like in the tests there won't be a `RepositorySignatureResource` or it won't specify `AllRepositorySigned`, so the downloader won't verify the signatures anyway and we shouldn't need all these test changes.

Forgind replied:
Short answer: yes.

You can see the full list of changes needed for the test in https://github.com/dotnet/sdk/pull/39268/commits/131cff5b873a66e4cf83c5048536c13e1781250e

You are generally correct that we won't verify signatures anyway in tests, but there are two tests that were different: the tests for unlisted packages. Since local feeds can't provide unlisted packages, JL03-Yue used the package the customer provided in writing tests to ensure that dotnet tool restore can find unlisted packages if necessary, and it turns out that package does claim to be signed, but the signature isn't valid, and it fails the test.

I have to admit that it took a lot of changes to thread through verifySignatures: false to a point where it could actually affect the test, and even after I got that part to work, the tests still needed more changes to find the right rid graph json, etc. If we don't want all those changes, the only alternative proposals I'm aware of are to either find another unlisted package that's signed correctly (but I/standup yesterday didn't know how to do that) or disable the test. marcpopMSFT sounded like he was most in favor of disabling the tests and avoiding all the other changes. If that's your preference as well, I can revert those last two commits and just disable them.

Forgind committed:
Comments

dtivel commented on src/Cli/dotnet/NugetPackageDownloader/LocalizableStrings.resx:
```diff
@@ -132,11 +132,18 @@
+    <value>Skipping signature verification for NuGet package "{0}" because it comes from a source that does not require signature validation.</value>
+  </data>
   <data name="SkipNuGetpackageSigningValidationmacOSLinux" xml:space="preserve">
     <value>Skip NuGet package signing validation. NuGet signing validation is not available on Linux or macOS https://aka.ms/workloadskippackagevalidation .</value>
```
This aka.ms link points to an announcement which doesn't apply to more recent versions of .NET.

Signed package verification is supported on Linux.  See https://learn.microsoft.com/en-us/dotnet/core/tools/nuget-signed-package-verification#linux.
